src/stdlib_io_filesystem.f90

```diff
@@ -0,0 +1,112 @@
+! SPDX-Identifier: MIT
+
+ !> Interaction with the filesystem.
+module stdlib_io_filesystem
+    use stdlib_string_type, only: string_type
+    implicit none
+    private
+
+    public :: exists, list_dir, run, temp_dir
+
+    character(*), parameter :: temp_dir = 'temp'
+    character(*), parameter :: listed_contents = temp_dir//'/listed_contents.txt'
+
+contains
+
+    !> Version: experimental
+    !>
+    !> Whether a file or directory exists at the given path.
+    !> [Specification](../page/specs/stdlib_io.html#exists)
+    logical function exists(filename)
+        !> Name of the file or directory.
+        character(len=*), intent(in) :: filename
+
+        inquire(file=filename, exist=exists)
+
+#if defined(__INTEL_COMPILER)
+        if (.not. exists) inquire(directory=filename, exist=exists)
+#endif
+    end
+
+    !> Version: experimental
+    !>
+    !> List files and directories of a directory. Does not list hidden files.
+    !> [Specification](../page/specs/stdlib_io.html#list_dir)
+    subroutine list_dir(dir, files, iostat, iomsg)
+        !> Directory to list.
+        character(len=*), intent(in) :: dir
+        !> List of files and directories.
+        type(string_type), allocatable, intent(out) :: files(:)
+        !> Status of listing.
+        integer, optional, intent(out) :: iostat
+        !> Error message.
+        character(len=:), allocatable, optional, intent(out) :: iomsg
+
+        integer :: unit, stat
+        character(len=256) :: line
+
+        stat = 0
+
+        if (.not. exists(temp_dir)) then
+            call run('mkdir '//temp_dir, stat)
```

Comment:
Note that under Windows OS, `mkdir temp` will create a `temp` folder under the current path `pwd`. Is this appropriate?

Reply:
It is a very pragmatic solution, but I think it should be fine for now